feat: add MagicEndpoint to iroh-net

Frando opened the pull request:
This is a first pass at adding a `MagicEndpoint` that combines a magicsock and a quinn endpoint. The API is minimalistic so far and allows to both accept and create connections. I also added two utility functions to extract the ALPN and PeerId on incoming connections, and a maximally-simple example to test things.

An endpoint-like struct that can create and accept connections will be needed for peer-to-peer sockets as e.g. in the upcoming gossipswarm impl. (It could also work with functions only but would involve more code copying I think).

I mostly copied code around from the existing listen and dial functions. No tests or other code is yet converted over to actually use this. Sharing now to get feedback on the API.

flub commented on iroh-net/src/endpoint.rs:
```diff
@@ -0,0 +1,245 @@
+            client_config
+        };
+
+        debug!("connecting to {}: (via {} - {:?})", peer_id, addr, addrs);
```
i wonder if the `addrs` is really correct here?  it could have even more addresses than we have here in the `addrs` variable i think?

Frando replied:
Very much possible, I copied this from `iroh-net/src/client.rs:132` and did not check it at all

flub commented on iroh-net/src/endpoint.rs:
```diff
@@ -0,0 +1,245 @@
+            .await
+            .expect("just inserted");
+
+        let client_config = {
+            let alpn_protocols = vec![alpn.to_vec()];
+            let tls_client_config =
+                tls::make_client_config(&self.keypair, Some(peer_id), alpn_protocols, self.keylog)?;
+            let mut client_config = quinn::ClientConfig::new(Arc::new(tls_client_config));
+            let mut transport_config = quinn::TransportConfig::default();
+            transport_config.keep_alive_interval(Some(Duration::from_secs(1)));
+            client_config.transport_config(Arc::new(transport_config));
+            client_config
+        };
```
This block could also be done at `quinn::Endpoint` creation time with `Endpoint::set_default_client_config`.  In this code-layout this doesn't matter and may even be harder to follow, but I'm not sure if wrapping the endpoint is the right approach.  See my top-level comment.

Frando replied:
Can it really? The crypto config contains the `peer_id` that you're connecting to.

flub replied:
oh, i missed that :(

flub reviewed (commented):
I've also been wondering about simplifying the use of the magicsocket.  But I was initially thinking of adding a bit of functionality to `Conn` (or maybe a new `MagicEndpoint, this is a pretty good name) that would have a constructor method to give you a `quinn::Endpoint` which is configured to use this magicsocket.

That way you'd have to call `.add_know_addrs()` and `.get_mapping_addr()` before you can call `.connect()` on a `quinn::Endpoint` making the relation between both sockets a bit more explicit.  But it would not put us in the game of having to wrap all `quinn::Endpoint` APIs.

Frando commented on iroh-net/src/defaults.rs:
```diff
@@ -0,0 +1,30 @@
+        host_name: "derp.iroh.network".into(),
+        stun_only: false,
+        stun_port: 3478,
+        ipv4: UseIpv4::Some("35.175.99.113".parse().unwrap()),
```
Ack to the first comment. Wrt to the second, I don't know ;-) Copied it from `iroh/config.rs`.

dignifiedquire replied:
> More interestingly though: are we fine with hardcoding an IP address here?

It is not great, but okay for the moment I would say

Frando commented:
> I've also been wondering about simplifying the use of the magicsocket. But I was initially thinking of adding a bit of functionality to `Conn` (or maybe a new `MagicEndpoint`, this is a pretty good name) that would have a method to give you a new `quinn::Endpoint` which is configured to use this magicsocket.

Aha, yeah that would work as well. Somehow, the notion of `connect(peer_id)` is quite appealing to me though.

> That way you'd have to call `.add_know_addrs()` and `.get_mapping_addr()` before you can call `.connect()` on a `quinn::Endpoint` making the relation between both ~sockets~ endpoints a bit more explicit. But it would not put us in the game of having to wrap all `quinn::Endpoint` APIs.

That is true. However, the API surface of `quinn::Endpoint` is really small when omitting the constructors which are taken care of in either case. Method-wise, it's just:
*  `accept`, `connect`: already implemented here
* `local_addr` is mostly useless unless coupled with `get_mapping_addr` I think? So the wrapping would make sense here IMO
* * `set_server_config` and `rebind`: not sure if exposing those in the magicsock setup is needed / makes sense
* `close` and `wait_idle`: those should be exposed

flub commented on iroh-net/src/endpoint.rs:
```diff
@@ -0,0 +1,439 @@
+        Ok(connection)
+    }
+
+    pub async fn add_known_addrs(
```
doc comment? (isn't that a lint?)

dignifiedquire replied:
the lint is currently disabled, there is a lot undocumented atm :(

flub replied:
ah, let's just keep adding issues to the milestone then :) https://github.com/n0-computer/iroh/issues/1139

dignifiedquire commented on iroh-net/examples/magic.rs:
```diff
@@ -0,0 +1,134 @@
+    tls::{Keypair, PeerId},
+};
+
+const EXAMPLE_ALPN: &[u8] = b"n0/iroh/examples/magic/0";
```
could be cool to make the alpn be also a cli argument

dignifiedquire commented on iroh-net/examples/magic.rs:
```diff
@@ -0,0 +1,134 @@
+        keypair,
+        args.bind_address,
+        vec![EXAMPLE_ALPN.to_vec()],
+        None,
```
these args get quite unwieldy, probably should be a builder struct

```rust
let endpoint = MagicEndpoint::builder()
  .keypair(keypair)
  .alpns(vec![EXAMPLE_ALPN.to_vec()])
  .derp_map(derp_map)
  .bind(args.bind_address).await?;
```

Frando commented:
Thanks for the reviews! I added a `MagicEndpointBuilder` and also more docs.

dignifiedquire commented:
@Frando would be great if you could do the integration/usage of it in `iroh` itself, so we can see how and if it works there as needed for the current use case.

dignifiedquire commented on iroh-net/src/endpoint.rs:
```diff
@@ -22,6 +22,78 @@ use crate::{
+    pub fn keypair(self, keypair: Keypair) -> Self {
+        Self {
+            keypair: Some(keypair),
+            ..self
```
interesting way to write this, any reason to use this over?

```rust
self.keypair.replace(keypair);
self
```

Frando replied:
no reason, changed it

dignifiedquire commented on iroh-net/src/endpoint.rs:
```diff
@@ -22,6 +22,78 @@ use crate::{
+    pub async fn bind(self, bind_addr: SocketAddr) -> anyhow::Result<MagicEndpoint> {
+        let keypair = self.keypair.unwrap_or_else(|| Keypair::generate());
+        let tls_server_config =
+            tls::make_server_config(&keypair, self.alpn_protocols, self.keylog)?;
```
does this error if `alpn_protocols` is empty?

Frando commented:
> @Frando would be great if you could do the integration/usage of it in `iroh` itself, so we can see how and if it works there as needed for the current use case.

Allright, I took a stab at this, the latest commit uses `MagicEndpoint` throughout `iroh`.

dignifiedquire commented:
> Allright, I took a stab at this, the latest commit uses `MagicEndpoint` throughout `iroh`.

Very nice and clean :) really like having this abstraction.

dignifiedquire commented:
@Frando the last integration I would love to see is replacing the structure in `hp/magicsock/conn.rs` tests

Frando commented:
> @Frando the last integration I would love to see is replacing the structure in `hp/magicsock/conn.rs` tests

I just had a look at this. The tests access more internals of the magic socket than the current API exposes. Most seems to me for validation of those internals. 
* they access the `tracked_endpoints()` of a magic socket to check if some peer id is in there.
* they also update the network map completely whereas the currently-exposed API on the `MagicEndpoint` only allows to `add_known_addr(peer_id, socket_addrs)`
I haven't worked with the internals of the magic socket much yet so not sure if those are needed outside of testing. So, should I ..
* expose `tracked_endpoints` and `set_network_map` on the `MagicEndpoint` or
* make the magic socket `Conn` on the MagicEndpint `pub(crate)` and keep tests mostly as they are
* restructure some of the tests to use less internals and work more with `PeerId`s only (likely possible, but not sure if the current setup is intended)

?

Frando commented:
And CI passed on the last commit, including CLI tests! Only netsim failures, but seems unrelated (github bucket something something)

dignifiedquire commented:
I think a first pass just exposing things that are needed using `cfg(test)` is fine. I can take a look in a second round if it makes to restructure the tests.

dignifiedquire commented on iroh-net/src/client.rs:
```diff
@@ -5,129 +5,37 @@ use std::{
+/// the remote endpoint, they can be specified and will be added to the endpoint's peer map.
+/// If no addresses are specified, the endpoint will try to dial the peer through the
+/// configured DERP servers.
 pub async fn dial_peer(
```
I think it would be nice to just make this a helper function on `MagicEndpoint`, eg `MagicEndpoint::dial_peer(..) -> Self`. That way we can reduce the API surface a little bit more, and all interactions just go through `MagicEndpoint`

Frando replied:
Yup, then we can remove the client module alltogether. The other function is only used in a single place (RPC client) and has not much to do with iroh_net as its just creating a quinn endpoint, so I'd just move that to where the RPC client is created for now.

dignifiedquire commented on iroh-net/src/endpoint.rs:
```diff
@@ -0,0 +1,542 @@
+use std::{
```
Can we call this `magic_endpoint.rs` please? I am always confused if the structs don't match the filename 😅

dignifiedquire commented on iroh-net/src/endpoint.rs:
```diff
@@ -0,0 +1,542 @@
+    /// make here is the address family: IPv4 or IPv6.  Otherwise you normally bind to the
+    /// `UNSPECIFIED` address on port `0` thus allowing the kernel to do the right thing.
+    ///
+    /// If *peer_id* is present it will verify during the TLS connection setup that the remote
```
not seeing `peer_id` as an argument

Frando replied:
Ah right, those were copied over from old docs. As the function is pub(crate) I'd move the docs to the builder.

dignifiedquire commented on iroh-net/src/endpoint.rs:
```diff
@@ -0,0 +1,542 @@
+        );
+        let connect = self
+            .endpoint
+            .connect_with(client_config, addr, "localhost")?;
```
Can you add a todo, to replace `localhost` with sth that makes more sense? (this is old, and okay for now, but should be checked/fixed eventually)

dignifiedquire reviewed (dismissed):
some smaller nits, other than that this is looking good :)

Frando commented:
I addressed the last comments and rebased onto main. Let's see what CI says.

One thing I'm not super sure about is the netmap managment. But I think this can be refined later.

dignifiedquire commented on iroh-net/examples/magic.rs:
```diff
@@ -0,0 +1,137 @@
+                    "https" => 443,
+                    _ => anyhow::bail!(
+                        "Invalid scheme in DERP URL, only http: and https: schemes are supported."
+                    ),
```
I have seen this logic in your other PR, maybe we can add a helper somewhere to keep this consistent?

Frando replied:
This would be obsoleted with https://github.com/n0-computer/iroh/pull/1143 I think because then we don't have to the port parsing here and just pass the URL through

dignifiedquire commented on iroh-net/src/hp/magicsock/conn.rs:
```diff
@@ -2649,7 +2658,7 @@ mod tests {
                         name: "t1".into(),
                         region_id: 1,
-                        host_name: "http://test-node.invalid".parse().unwrap(),
+                        host_name: "https://test-node.invalid".parse().unwrap(),
```
does this need to be `https`?

dignifiedquire commented on iroh-net/src/lib.rs:
```diff
@@ -1,7 +1,10 @@
-pub mod client;
+pub mod defaults;
 pub mod hp;
+pub mod magic_endpoint;
```
can we make this private, if we reexport the `MagicEndpoint`?

Frando replied:
There's also the `MagicEndpointBuilder` struct and `accept_conn` functions in there (and possible `get_alpn` and `get_peer_id`, see below). Not sure if they should all be at the top level?
I think I'd defer this to later, currently pretty much everything in `iroh-net` is public anyway - and once things stabilize a bit we might want to change this overall and actually define the public API surface for `iroh-net`?

dignifiedquire commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,562 @@
+        callbacks: Option<Callbacks>,
+        keylog: bool,
+    ) -> anyhow::Result<Self> {
+        // let (endpoints_update_s, endpoints_update_r) = flume::bounded(1);
```
still needed?

dignifiedquire commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,562 @@
+}
+
+/// Extract the [`PeerId`] from the peer's TLS certificate.
+pub async fn get_peer_id(connection: &quinn::Connection) -> anyhow::Result<PeerId> {
```
does this need to be public?

dignifiedquire replied:
lets leave it as is for now then

dignifiedquire commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,562 @@
+}
+
+/// Extract the ALPN protocol from the peer's TLS certificate.
+pub async fn get_alpn(connecting: &mut quinn::Connecting) -> anyhow::Result<String> {
```
does this need to be public?

dignifiedquire reviewed (dismissed):
some nits, other than that, lgtm

flub commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,561 @@
+        })
+    }
+
+    /// Accept a connection on the socket.
```
```suggestion
    /// Accept an incoming connection on the socket.
```

flub commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,561 @@
+}
+
+/// Accept an incoming connection and extract the client-provided [`PeerId`] and ALPN protocol.
+pub async fn accept_conn(
```
I'd be tempted to experiment with an extension trait or a wrapper for `Connecting` (and I guess `Accept`).  But we can do that later.

Frando replied:
Yup I thought about that too. However we'd either have to use `async_trait` or write the futures by hand. I'd defer this to post-merge I think.

flub commented on iroh-net/src/hp/magicsock/conn.rs:
```diff
@@ -2826,7 +2820,7 @@ mod tests {
     }
 
-    fn setup_logging() {
+    pub fn setup_logging() {
```
I kind of wish you copied in `iroh-bytes/src/test_utils.rs` to `iroh-net/src/test_utils.rs` and used it's `setup_logging` call for your functions instead of making this pub...  But I tried to clean this up earlier and utterly failed failed.  For new `#[tokio::test]` tests it should just work though, since they're single-threaded runtimes.  Just don't try change existing tests.

But if you don't want to do that now that's fine, it's my fault for not managing to clean this up before.

Frando replied:
Yes this sounds good to me too. I also want to expose `run_derp_and_stun` in a test_util module also to other crates (i.e. iroh-gossip) that use iroh-net and need to test things with a local derper. 
However I'd prefer to do this after the merge in a new PR.

flub reviewed (approved):
Nice!  Most of my comments are about the public docs from a user point of view.  But nothing is blocking.

Frando commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,561 @@
+    /// The *bind_addr* is the address that should be bound locally.  Even though this is an
+    /// outgoing connection a socket must be bound and this is explicit.  The main choice to
+    /// make here is the address family: IPv4 or IPv6.  Otherwise you normally bind to the
+    /// `UNSPECIFIED` address on port `0` thus allowing the kernel to do the right thing.
```
Good question, this is taken over from the old `dial_peer` functions and the code in `iroh/src/node.rs`
Now that I re-read the code, the `bind_addr` IP address & family  is not even used, only its port is passed to the `magicsock::Conn`. So - it would actually be correct to remove the `bind_addr: SocketAddr` argument and instead add `bind_port: u16`?
@dignifiedquire can you confirm this? Then I'd go ahead and replace `bind_addr` with `bind_port`.

dignifiedquire replied:
My answer is two fold, the current state and the ideal state.

- Current State: The address is effectively ignored and only the port is used, as we bind to IPv4 and IPv6 seperately.
- Should State: Allow to provide sth like this 
```rust
enum BindAddress {
  /// Bind on this port, both ipv4 and ipv6 addr, using `0.0.0.0` and `[::]`
  Port(u16),
  Ips(Ips),
}
enum Ips {
  // Only Ipv4
  Ipv4(SocketV4Addr),
  // Only Ipv6
  Ipv6(SocketV6Addr),
  Both {
    ipv4: SocketV4Addr,
    ipv6: SocketV4Addr,
}

// The `BindAddress::Port` options is equivalent to 
BindAddress::Ips(Ips::Both {
  ipv4: SocketV4Addr::new(Ipv4Addr::UNSPECIFIED, port),
  ipv6: SocketV6Addr::new(Ipv6Addr::UNSPECIFIED, port),
});

// so could probably change this to just have `Ips` and this
impl Ips {
  pub fn from_port(port: u16) -> Self {
    Ips::Both {
      ipv4: SocketV4Addr::new(Ipv4Addr::UNSPECIFIED, port),
      ipv6: SocketV6Addr::new(Ipv6Addr::UNSPECIFIED, port),
    }
  }
}
```

flub replied:
Yes, that makes more sense.  Maybe `Ips` should even be a vec, I think there's nothing wrong with binding to a bunch of specific addresses, e.g. you're on say 20 subnets but only want to bind to the private ones.

flub replied:
So I'd be mildly in favour of for now changing this to `u16` and describe that this will change in the future to allow binding to specific socket addresses.

dignifiedquire replied:
> e.g. you're on say 20 subnets but only want to bind to the private ones

That would be a lot of internal changes, as we only do at most two sockets internally. I would argue, this should be managed by some network configuration/proxy infront if you have multiple private binds.

Frando commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,561 @@
+    /// Get the local endpoints on which the underlying magic socket is listening.
+    ///
+    /// This returns information on the network state of the endpoints in addition to
+    /// their addresses.
```
Good question! I was thinking the `local_endpoints` would also contain addresses obtained via STUN. Is this correct?

If this is the case, then maybe something like this?

```
Get the local and discovered endpoints on which the underlying magic socket is reachable.

This list contains both the locally-bound addresses and the endpoint's publicly-reachable addresses, if they could be discovered through STUN or port mapping.
```

dignifiedquire replied:
Yes, the endpoints are all that we have discovered about ourselves

Frando commented on iroh-net/src/magic_endpoint.rs:
```diff
@@ -0,0 +1,561 @@
+    /// Connect to a remote endpoint, creating an endpoint on the fly.
+    ///
+    /// The PeerId and the ALPN protocol are required. If you happen to know dialable addresses of
+    /// the remote endpoint, they can be specified and will be added to the endpoint's peer map.
```
I removed the term and took over the description you proposed in `connect`
